sphinx/testing/matcher/__init__.py

```diff
@@ -0,0 +1,365 @@
+        """Temporarily extend the set of options with *options*."""
+        self.__stack.append(None)  # prepare the next cache entry
+        try:
+            with super().override(**options):
+                yield
+        finally:
+            self.__stack.pop()  # pop the cached lines
+
+    @property
+    def content(self) -> str:
+        """The raw content."""
+        return self.__content
+
+    def lines(self) -> Block:
+        """The content lines, cleaned up according to the current options.
+
+        This method is efficient in the sense that the lines are computed
+        once per set of options and cached for subsequent calls.
+        """
+        stack = self.__stack
+        assert stack, 'invalid stack state'
+        cached = stack[-1]
+
+        if cached is not None:
+            if isinstance(cached, int):
+                return cast(Block, self.__stack[cached])
+            return cached
+
+        lines = self.__get_clean_lines()
+        # check if the value is the same as any of a previously cached value
+        # but do not use slices to avoid a copy of the stack
+        for addr, value in zip(range(len(stack) - 1), stack):
+            if isinstance(value, int):
+                cached = cast(Block, stack[value])
+                if cached.buffer == lines:
+                    # compare only the lines as strings
+                    stack[-1] = value  # indirection near to beginning
+                    return cached
+
+            if isinstance(value, Block):
+                if value.buffer == lines:
+                    stack[-1] = addr  # indirection
+                    return value
+
+        # the value did not exist yet, so we store it at most once
+        stack[-1] = cached = Block(lines, _check=False)
+        return cached
+
+    def find(
+        self,
+        patterns: LinePattern | Set[LinePattern] | Sequence[LinePattern],
+        /,
+        *,
+        flavor: Flavor | None = None,
+    ) -> tuple[Line, ...]:
+        """Same as :meth:`iterfind` but returns a sequence of lines."""
+        # use tuple to preserve immutability
+        return tuple(self.iterfind(patterns, flavor=flavor))
+
+    def iterfind(
+        self,
+        patterns: LinePattern | Set[LinePattern] | Sequence[LinePattern],
+        /,
+        *,
+        flavor: Flavor | None = None,
+    ) -> Iterator[Line]:
+        """Yield the lines that match one (or more) of the given patterns.
+
+        :param patterns: The patterns deciding whether a line is selected.
+        :param flavor: Optional temporary flavor for non-compiled patterns.
+
+        By convention, the following are equivalent::
+
+            matcher.iterfind('line to find', ...)
+            matcher.iterfind(['line to find'], ...)
+        """
+        patterns = _engine.to_line_patterns(patterns)
+        if not patterns:  # nothinig to match
+            return
+
+        compiled_patterns = set(self.__compile(patterns, flavor=flavor))
```

Comment:
What's the benefit in de-duplicating the compiled patterns?

Reply:
Efficiency, just in case. Since it's likely to be used in some testing code, I predict that there might be copy-paste here and there. But technically, I can ignore that and just say "you should know what to do".

Reply:
Ok, I agree with the efficiency argument, and would also note that de-duplicating regexes shouldn't ever cause change-in-behaviour - if `x` matches `y`, it'll always match, whether it's checked once or ten times -- so let's check once.

But re-ordering the order in which matches are applied _could_ be hard-to-debug; so I think I'd prefer an order-preserving dedup here if that's OK with you?

(to explain: a human is likely to order regexes based on how likely they think they are to catch problems -- and it's possible that they could re-arrange those to be performance-optimized, whereas a Python `set` could disrupt that ordering)

Reply:
Ah, fine by me. I have a utility function for that (it's the unique_everseen one)